ext/GDBM_File/GDBM_File.xs

```diff
@@ -62,17 +145,114 @@ output_datum(pTHX_ SV *arg, char *str, int size)
 #define gdbm_setopt(db,optflag,optval,optlen) not_here("gdbm_setopt")
 #endif
 
+#ifndef GDBM_ITEM_NOT_FOUND
+# define GDBM_ITEM_NOT_FOUND GDBM_NO_ERROR
+#endif
+
+/* Prior to 1.13, gdbm_fetch family functions set gdbm_errno to GDBM_NO_ERROR
+   if the requested key did not exist */
+#define ITEM_NOT_FOUND()                                                \
+    (gdbm_errno == GDBM_ITEM_NOT_FOUND || gdbm_errno == GDBM_NO_ERROR)
+
+#define CHECKDB(db) do {                        \
+    if (!db->dbp) {                             \
+        croak("database was closed");           \
+    }                                           \
+ } while (0)
+
 static void
-croak_string(const char *message) {
-    Perl_croak_nocontext("%s", message);
+dbcroak(GDBM_File db, char const *func)
+{
+#if GDBM_VERSION_MAJOR == 1 && GDBM_VERSION_MINOR >= 13        
+    croak("%s: %s", func, gdbm_db_strerror(db->dbp));
+#else
+    (void)db;
+    croak("%s: %s", func, gdbm_strerror(gdbm_errno));
+#endif
+}
+
+#if GDBM_VERSION_MAJOR == 1 && (GDBM_VERSION_MINOR > 16 || GDBM_VERSION_PATCH >= 90)
+# define gdbm_close(db)    gdbm_close(db->dbp)
+#else
+# define gdbm_close(db)    (gdbm_close(db->dbp),0)
+#endif
+static int
+gdbm_file_close(GDBM_File db)
+{
+    int rc = 0;
+    if (db->dbp) {
+        rc = gdbm_close(db);
+        db->dbp = NULL;
+    }
+    return rc;
+}
+
+#if GDBM_VERSION_MAJOR == 1 && GDBM_VERSION_MINOR >= 13
+/* Error-reporting wrapper for gdbm_recover */
+static void
+rcvr_errfun(void *cv, char const *fmt, ...)
+{
```

Comment:
This needs a dTHX to build with threaded builds of perl:

Reply:
Fixed in new commit. Thank you.